cmd/machine-config-daemon/start.go

```diff
@@ -69,22 +67,44 @@ func runStartCmd(cmd *cobra.Command, args []string) {
 
 	stopCh := make(chan struct{})
 	defer close(stopCh)
-
-	ctx := common.CreateControllerContext(cb, stopCh, componentName)
-	// create the daemon instance. this also initializes kube client items
-	// which need to come from the container and not the chroot.
-	daemon, err := daemon.New(
-		startOpts.rootMount,
-		startOpts.nodeName,
-		operatingSystem,
-		daemon.NewNodeUpdaterClient(),
-		cb.MachineConfigClientOrDie(componentName),
-		cb.KubeClientOrDie(componentName),
-		daemon.NewFileSystemClient(),
-		ctx.KubeInformerFactory.Core().V1().Nodes(),
-	)
-	if err != nil {
-		glog.Fatalf("failed to initialize daemon: %v", err)
+	var dn *daemon.Daemon
+
+	// If we are asked to run once and it's a valid file system path use
+	// the bare Daemon
+	if startOpts.onceFrom != "" && daemon.ValidPath(startOpts.onceFrom) {
+		dn, err = daemon.New(
+			startOpts.rootMount,
+			startOpts.nodeName,
+			operatingSystem,
+			daemon.NewNodeUpdaterClient(),
+			daemon.NewFileSystemClient(),
+			startOpts.onceFrom,
+		)
+		if err != nil {
+			glog.Fatalf("failed to initialize single run daemon: %v", err)
+		}
+		// Else we use the cluster driven daemon
+	} else {
+		// create the daemon instance. this also initializes kube client items
+		// which need to come from the container and not the chroot.
+		dn, err = daemon.NewClusterDrivenDaemon(
+			startOpts.rootMount,
+			startOpts.nodeName,
+			operatingSystem,
+			daemon.NewNodeUpdaterClient(),
+			startOpts.kubeconfig,
+			daemon.NewFileSystemClient(),
+			startOpts.onceFrom,
+			stopCh,
+			componentName,
+		)
+		if err != nil {
+			glog.Fatalf("failed to initialize daemon: %v", err)
+		}
+		err = dn.CheckStateOnBoot(stopCh)
```

Comment:
So now that you start the informer in `NewClusterDrivenDaemon` above, we run into the problem that is: let's say a previous run degraded the node while updating, so we never got to the state where `desiredConfig` = `currentConfig`. Upon restarting, we start the informer and it will realize we need an update (dc != cc), and trigger that likely before we reach the CheckStateOnBoot (since in practice the update gets triggered quite often from each node). This way instead of sleeping (which preserves the actual degrade error from the previous run), we crash loop instead.

I recommend splitting out the starting of the informer to after this line instead.

Reply:
Updated with logic we came to via chat.

Reply:
Confirmed, looks good!